ambry-store/src/main/java/com/github/ambry/store/StorageManager.java

```diff
@@ -476,16 +477,40 @@ public void onPartitionBecomeBootstrapFromOffline(String partitionName) {
         // able to get new replica from storageManager without querying Helix
       } else {
         // if the replica is already on current node, there are 3 cases need to discuss:
-        // 1. replica was initially present in clustermap;
-        // 2. replica was dynamically added to this node but may fail during BOOTSTRAP -> STANDBY transition
+        // 1. replica was initially present in clustermap and this is a regular reboot.
+        // 2. replica was dynamically added to this node but may fail during BOOTSTRAP -> STANDBY transition.
         // 3. replica is on current node but its disk is offline. The replica is not able to start.
+        // 4. replica was initially present in clustermap but it's current being recreated due to disk failure before.
+
         // For case 1 and 2, OFFLINE -> BOOTSTRAP is complete, we leave remaining actions (if there any) to other transition.
         // For case 3, we should throw exception to make replica stay in ERROR state (thus, frontends won't pick this replica)
-        if (getStore(replica.getPartitionId(), false) == null) {
+        // For case 4, we check it's current used capacity and put it in BOOTSTRAP state if necessary. This is to ensure
+        //             it catches up with peers before serving PUT traffic (or being selected as LEADER)
+        Store store = getStore(replica.getPartitionId(), false);
+        if (store == null) {
           throw new StateTransitionException(
               "Store " + partitionName + " didn't start correctly, replica should be set to ERROR state",
               StoreNotStarted);
         }
+        // if store's used capacity is less than or equal to header size, we create a bootstrap_in_progress file and force
+        // it to stay in BOOTSTRAP state when catching up with peers.
+        long storeUsedCapacity = store.getSizeInBytes();
+        if (storeUsedCapacity <= HEADER_SIZE) {
```

Comment:
I wonder if this is really necessary. If we create a bootstrap_in_progress file no matter the used capacity and let the remaining transition remove it later. In this way I think we can address another corner case where only half of the data in this partition gets wiped out.

Imagine that we have two log segment files in this partition, somehow we remove the second log segment file and then we restart the process. Here since the used capacity is no less then HEADER_SIZE, so the bootstrap_in_progress file will not be created, then the replicationmanager's callback would just mark it as standby on the onPartitionBecomeStandByFromBootstrap.

But if we create the bootstrap without checking the HEADER_SIZE, then we can avoid this issue.

Reply:
We can extend this to some other interesting edge cases, I will talk with you offline. For now, let's start with empty store first, that's the most common case when we wipe out whole partition or replace the disk.